src/coprocessor/dag/executor/metrics.rs

```diff
@@ -113,3 +113,160 @@ impl ExecCounter {
             .inc_by(self.aggregation);
     }
 }
+
+/// Collector that collects the number of scanned keys of each range.
+pub trait CountCollector {
+    /// Notifies the collector that a new range is started.
+    fn handle_new_range(&mut self);
+
+    /// Increases the counting of number of scanned keys of current range. If no range is started
+    /// this function has no effect.
+    fn inc_counter(&mut self);
+
+    /// Takes and appends number of scanned keys into `target`.
+    fn collect(&mut self, _target: &mut Vec<i64>);
+}
+
+/// A normal `CountCollector`.
+#[derive(Default)]
+pub struct CountCollectorNormal {
+    counts: Vec<i64>,
+}
+
+impl CountCollector for CountCollectorNormal {
+    #[inline]
+    fn handle_new_range(&mut self) {
+        self.counts.push(0);
+    }
+
+    #[inline]
+    fn inc_counter(&mut self) {
+        self.counts.last_mut().map_or((), |val| *val += 1);
+    }
+
+    #[inline]
+    fn collect(&mut self, target: &mut Vec<i64>) {
+        target.append(&mut self.counts);
+        self.counts.push(0);
+    }
+}
+
+/// A `CountCollector` that does not collect anything.
+pub struct CountCollectorDisabled;
+
+impl CountCollector for CountCollectorDisabled {
+    #[inline]
+    fn handle_new_range(&mut self) {}
```

Comment:
We could, but `ExecutionSummaryCollector` could not (because there is an associate type inside and a function returning the associate type; the function returning the associate type cannot have default implementation due to compiler restrictions. the compiler expects the default function to return the associate type instead of the default type of the associate type). Believe me that initially I was writing the code just like you suggested, and later found out the facts above so changed them to this more verbose form.

Reply:
So as `ExecutionSummaryCollector` has a type and must return the associated type. Do we do this in order to make the code style keep consistent?

Reply:
I prefer to keep consistent so that we don't need further comment to explain why something can't be done here 🤣 Anyway this is rather trivial and I guess even in future only Normal and Disabled is possible so that default implementation doesn't have a notable advantage (it only serves for Disabled).